### ISSUE TITLE
special-lispy-tab deletes part of let-block (Racket Mode)

### ISSUE THREAD

**Issue by BrandonIrizarry** [abo-abo/lispy#430]
Say I have 
```
(let ((a 4)
      (b 6))
  (+ a b))
```
If I apply `i`, that is, `special-lispy-tab`, at the end of the variable-declaration list, I get
```
(let )
(b 6))
(+ a b))
```
Also, say I have
```
(module+ test
  (test-case
      "blah"
    (check-equal? 3 3)
    (check-equal? 4 4)))
```

After applying `special-lispy-tab` at the end of the `test-case` form, I get
```
(module+ test
  est-case
  "blah"
  (check-equal? 3 3)
  (check-equal? 4 4)))
```
This, of course, leads to unbalanced parentheses, and is very confusing!  I think that this has something to do with "outlines", but then again, I'm not sure about that.

**Comment by BrandonIrizarry** [comment]
OK - I got `lispy` working, under emacs -Q.

In Fundamental Mode, the let-example (as you pointed out) doesn't break.  However, I'm using[ racket-mode   ](https://github.com/greghendershott/racket-mode) for editing Racket code, and I notice that the indentation-style that `special-lispy-tab` applies is different from the convention that Racket Mode uses (which I prefer.)

When I also install `racket-mode` alongside `lispy` under emacs -Q, and set `racket-mode` as my major-mode for editing Racket code, first, I notice that` lispy` is actually disabled upon switching to this new mode.  After reenabling `lispy`, I notice the behavior I initially reported.

I also tried `scheme-mode` and `geiser-mode` (under my normal Emacs),  and yes, the examples don't break (although, again, the indentation convention for submodules isn't ideal for Racket, in my opinion.)  I suspect, then, that this has something to do with the peculiar way in which `racket-mode` is implemented.

**Comment by noctuid** [comment]
I cannot replicate the issue in `racket-mode`.

**Comment by abo-abo** [comment]
> with the peculiar way in which racket-mode is implemented

This is likely the case. Likely either `indent-sexp` or `forward-sexp` is broken in your example in `racket-mode`.

**Comment by bestlem** [comment]
Actually it is not indent-sexp that is broken the issue might be that lispy does not call indent-sexp.
 lispy--normalize-1 only calls indent-sexp if the size is more than 10000 characters or you are in a clojure mode and some specific syntax.

The issue looks to me to be that lispy--prin1-to-string is not doing the correct thing in racket mode. (My lisp experience is not enough to work out what is happening there).

I altered   lispy--normalize-1 to call indent-sexp (although the author or racket suggests it should be prog-indent-sexp https://github.com/greghendershott/racket-mode/blob/master/racket-indent.el) and on a few cases seems to work except it does not remove all the spaces between end parentheses.

My test case was 
 (define x 1)
and it should remove the beginning space but lispy mode removed 2 characters to leave as efine x 1)

**Comment by abo-abo** [comment]
Still can't reproduce neither the original issue nor the `(define x 1)` case with racket-mode-20190607.109.

Please try to provide a reproducible scenario.

**Comment by bestlem** [comment]
To make it clearer my (define x 1) had a space before the opening bracket.

**Comment by abo-abo** [comment]
OK, I've narrowed down the reason to a bug in `racket-mode`.

```racket
__(define x 1)
```

Open a new `racket-mode` buffer, and insert an expression indented by two spaces.
Navigate to the expression start and call `indent-sexp`.

- Expected: the buffer is not changed. This is the behavior of all other LISP-like modes.
- Actual: the leading spaces are removed.

**Comment by abo-abo** [comment]
I've added a workaround for `racket-mode`. It's ugly if there's an offset, but at least the expression stays balanced. Please report a bug for `racket-mode`.

**Comment by bestlem** [comment]
@Abo-abo I am not certain that it is only racket mode here. The non working code goes through lispy--normalize-1 which only uses   (indent-sexp) on long strings or for some clojure syntax.

In my fork I have added a test that shows the reported problem when run before your workaround and also shows a misindented but valid syntax after your fix.

(should (string= (lispy-with racket "  (defun test? (x) x)|" "i")
                   "(defun test? (x) x)|")) ; this requires a (require 'racket-mode)

I have tried altering lispy--normalize-1 to use indent-sexp but it gives the same error as your fix.

**Comment by bestlem** [comment]
As a note
I created a fork with ert tests for this case

(should (string= (lispy-with racket "  (defun test?  (x) x)|" "i")
                   "(defun test? (x) x)|"))

It fails in the original case with losing the first 2 characters, with our fix it does not indent correctly as you say but is valid.
I tried another fix of using indent-sexp in lispy--normalise-1 as you do for long strings or some clojure syntax - that fixed the indentation but does collapse multiple spaces down to one.
